simple_transformers/transformer.py

```diff
@@ -0,0 +1,105 @@
+import numpy as np
+import torch as th
+import torch.nn as nn
+from simple_transformers.modality_processors import MODALITY_PROCESSORS
+from simple_transformers.utils import _init_weights, get_config
+
+from typing import Any, Union, List, Tuple
+
+
+class ModalityEncoder(nn.Module):
+    def __init__(self, modality: str, **kwargs):
+        """
+        Encode most modalities using a transformer
+        :param modality: A string defining kind of modality to encode. e.g. "text" or "images".
+                         See MODALITY_PROCESSORS in simple_transformers.modality_processors for all options
+        :param num_classes: int or None. If int, creates a linear layer to output class logits
+        :param kwargs: kwargs that define the data. Requirements vary by dataset and modality
+        """
+        super().__init__()
+        self.config = get_config()
+        if modality not in MODALITY_PROCESSORS:
+            raise NotImplementedError(f'A processor for {modality} has yet to be implemented.'
+                                      f'The following modalities do have processors: {MODALITY_PROCESSORS.keys()}.')
+        self.preprocessor = MODALITY_PROCESSORS[modality](self.config, **kwargs)
+        self.encoder_layers = nn.TransformerEncoderLayer(self.config.d_model, self.config.n_heads,
+                                                         self.config.hidden_size, self.config.dropout_prob,
+                                                         batch_first=True)
+        self.transformer_encoder = nn.TransformerEncoder(self.encoder_layers, self.config.n_layers)
+        self.num_classes = None
+        if 'num_classes' in kwargs:
+            self.num_classes = kwargs['num_classes']
+            self.classifier = nn.Linear(self.config.d_model, self.num_classes)
+        self.apply(_init_weights)
+        self.to(self.config.device)
+
+    def forward(self, model_input: Any, attention_mask: Union[np.array, None] = None,
+                return_full_output=False) -> th.Tensor:
+        """
+        :param model_input: input modality. Input type depends on the modality being encoded (e.g. for text, use str)
+        :param attention_mask: Attention mask on input modality. None if attention mask is dealt using modalit processor
+        :return: Depending on if num classes has been defined and return_full_output, return either:
+                 1) Full output of the transformer (all token embds). Shape = (batch size, seq len, d_model)
+                 2) Single embedding for the full input (using the CLS token emb). Shape = (batch size, d_model)
+                 3) If num classes has been defined, return logits. Shape = (batch size, num_classes)
```

Comment:
Good question. Not sure yet what the most common usage will be. My primary usage will be to get the embeddings. They are in different spots because to classifying things, it requires initializing things (i.e. a linear layer which has to be done in init), while full_embeddings or not is easy to change on the fly